src/cata_tiles.cpp

```diff
@@ -972,51 +962,156 @@ void cata_tiles::draw_rhombus(int destx, int desty, int size, SDL_Color color, i
+    // Center the drawn area within the total area.
+    minimap_drawn_width = minimap_tiles_x_limit * minimap_tile_size_x;
+    minimap_drawn_height = minimap_tiles_y_limit * minimap_tile_size_y;
+    minimap_border_width = std::max((width - minimap_drawn_width) / 2, 0);
+    minimap_border_height = std::max((height - minimap_drawn_height) / 2, 0);
+    //prepare the minimap clipped area
+    minimap_clip_rect.x = destx + minimap_border_width;
+    minimap_clip_rect.y = desty + minimap_border_height;
+    minimap_clip_rect.w = width - minimap_border_width * 2;
+    minimap_clip_rect.h = height - minimap_border_height * 2;
+}
+
+//the main call for drawing the pixel minimap to the screen
 void cata_tiles::draw_minimap( int destx, int desty, const tripoint &center, int width, int height )
 {
     if (!g) {
         return;
     }
 
-    const int minx = 0;
-    const int miny = 0;
-    const int maxx = MAPSIZE * SEEX;
-    const int maxy = MAPSIZE * SEEY;
+    //set up class variables on the first run
+    if(!minimap_prep) {
+        init_minimap( destx, desty, width, height );
+    }
 
-    const int tiles_range_x = (MAPSIZE - 2) * SEEX;
-    const int tiles_range_y = (MAPSIZE - 2) * SEEY;
-    const int tile_size_x = std::max(width / tiles_range_x, 1);
-    const int tile_size_y = std::max(height / tiles_range_y, 1);
-    const int tiles_x_limit = std::min(width / tile_size_x, tiles_range_x);
-    const int tiles_y_limit = std::min(height / tile_size_y, tiles_range_y);
-    const int start_x = center.x - tiles_x_limit / 2;
-    const int start_y = center.y - tiles_y_limit / 2;
+    //clear leftover flags for the current draw cycle
+    prepare_minimap_cache_for_updates();
 
-    // Center the drawn area within the total area.
-    const int drawn_width = tiles_x_limit * tile_size_x;
-    const int drawn_height = tiles_y_limit * tile_size_y;
-    const int border_width = std::max((width - drawn_width) / 2, 0);
-    const int border_height = std::max((height - drawn_height) / 2, 0);
+    const int start_x = center.x - minimap_tiles_x_limit / 2;
+    const int start_y = center.y - minimap_tiles_y_limit / 2;
 
     auto &ch = g->m.access_cache( center.z );
+    //retrieve night vision goggle status once per draw
+    auto vision_cache = g->u.get_vision_modes();
+    bool nv_goggle = vision_cache[NV_GOGGLES];
 
-    SDL_Rect rectangle;
-    rectangle.w = tile_size_x;
-    rectangle.h = tile_size_y;
 
-    // First draw terrain.
-    for( int y = 0; y < tiles_y_limit; y++) {
-        if(start_y + y < miny || start_y + y >= maxy){
-            continue;
-        }
-        for( int x = 0; x <= tiles_x_limit; x++) {
-            if(start_x + x < minx || start_x + x >= maxx){
-                continue;
-            }
-            tripoint p(start_x + x, start_y + y, center.z);
+    //check all of exposed submaps (MAPSIZE*MAPSIZE submaps) and apply new color changes to the cache
+    for( int y = 0; y < MAPSIZE * SEEY - 1; y++) {
```

Comment:
That was actually an off-by-one leftover from testing, one run crashed because of the <= touched MAPSIZE * SEEX outside of the array, and I thought both variables were using <=.